OpenRA.Mods.Common/Traits/Attack/AttackFrontal.cs

```diff
@@ -40,7 +40,7 @@ protected override bool CanAttack(Actor self, Target target)
 
 			var f = facing.Facing;
 			var pos = self.CenterPosition;
-			var targetedPosition = Info.AttackTargetCenter ? target.CenterPosition : target.Positions.PositionClosestTo(pos);
+			var targetedPosition = HasAnyValidWeapons(target, true) ? target.CenterPosition : target.Positions.PositionClosestTo(pos);
```

Comment:
Could we define an `AttackBase.GetTargetPosition` instead of repeating this line in all the subclasses?